File: src/test/java/com/mkreidl/ephemeris/solarsystem/MoonPhaseTest.java
package com.mkreidl.ephemeris.solarsystem;

import com.mkreidl.ephemeris.Position;
import com.mkreidl.ephemeris.TestUtil;
import com.mkreidl.ephemeris.TestUtil.EphemerisData;
import com.mkreidl.ephemeris.Time;
import com.mkreidl.ephemeris.geometry.Angle;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import java.util.Collections;

import static com.mkreidl.ephemeris.solarsystem.Body.EARTH;
import static com.mkreidl.ephemeris.solarsystem.Body.MOON;

@RunWith( Parameterized.class )
public class MoonPhaseTest
{
    private final SolarSystem solarSystem = new SolarSystemMeeus();
    private final Time time;
    private final Body body;
    private final EphemerisData expected;

    public MoonPhaseTest( String testname, Body body, Time time, EphemerisData expected )
    {
        this.body = body;
        this.time = time;
        this.expected = expected;
    }

    @Parameters( name = "{0}" )
    public static Iterable<Object[]> data()
    {
        return TestUtil.solarSystemData( Collections.singletonList( Body.MOON ) );
    }

    @Test
    public void testMoonPhase()
    {
        if ( body == MOON )
        {
            solarSystem.compute( time, MOON );
            solarSystem.compute( time, EARTH );
            final Position actual = solarSystem.getEphemerides( MOON, new Position() );
            // expected.phase has an accuracy to only 0.36, since input data (accuracy 1e-3) was multiplied with 360°
            System.out.println( actual.getPhase( new Angle() ).get( Angle.Unit.DEGREES ) );
            System.out.println( actual.getIlluminatedFraction() );
            System.out.println( ( expected.phase + 180 ) / 360 );
            System.out.println( "=====================" );
            //
            // Test omitted, since reference seems incorrect
            //assertEquals( expected.phase, actual.getPhase( new Angle() ).get( Angle.Unit.DEGREES ), 0.36 );
        }
    }
}

File: src/test/java/com/mkreidl/ephemeris/solarsystem/MoonPhaseTest.kt
package com.mkreidl.ephemeris.solarsystem

import com.mkreidl.ephemeris.time.Instant
import com.mkreidl.ephemeris.TestUtil
import com.mkreidl.ephemeris.TestUtil.EphemerisData
import com.mkreidl.ephemeris.solarsystem.Body.MOON
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters

@RunWith(Parameterized::class)
class MoonPhaseTest(testname: String, private val body: Body, private val instant: Instant, private val expected: EphemerisData) {
    private val solarSystem = SolarSystem.createFromMeeus()

    @Test
    fun testMoonPhase() {
        if (body === MOON) {
            solarSystem.compute(instant)
            // expected.phase has an accuracy to only 0.36, since input data (accuracy 1e-3) was multiplied with 360°
            println(solarSystem.getPhase(MOON).degreesNonNeg)
            println(solarSystem.getIlluminatedFraction(MOON))
            println((expected.phase + 180) / 360)
            println("=====================")
            //
            // Test omitted, since reference seems incorrect
            //assertEquals( expected.phase, actual.getPhase( new Angle() ).get( Angle.Unit.DEGREES ), 0.36 );
        }
    }

    companion object {
        @JvmStatic
        @Parameters(name = "{0}")
        fun data() = TestUtil.solarSystemData(listOf(Body.MOON))
    }
}
